### PR TITLE
[MNG-7406] Better display of project building errors

### PR DESCRIPTION
**Pull request by mthmulders**
Following this checklist to help us incorporate your
contribution quickly and easily:

 - [X] Make sure there is a [MNG-7406](https://issues.apache.org/jira/browse/MNG-7406) filed
       for the change (usually before you start working on it).  Trivial changes like typos do not
       require a JIRA issue. Your pull request should address just this issue, without
       pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[MNG-XXX] SUMMARY`, where you replace `MNG-XXX`
       and `SUMMARY` with the appropriate JIRA issue. Best practice is to use the JIRA issue
       title in the pull request title and in the first line of the commit message.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will
       be performed on your pull request automatically.
 - [X] You have run the [Core IT][core-its] successfully - running them on GitHub Actions as a line of first defence.

ICLA and CCLA signed.

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Comment by michael-o** [comment]
Interesting, will happily review next couple of days...

File: maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java [hunk 1]
```diff
@@ -72,7 +67,7 @@ protected ProjectBuildingException( String projectId, String message, File pomFi
 
     public ProjectBuildingException( List<ProjectBuildingResult> results )
     {
-        super( createMessage( results ) );
+        super( "Some problems were encountered while processing the POMs" );
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java:70]: POMs? Not POM?

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuilderTest.java [hunk 2]
```diff
@@ -103,9 +106,9 @@ public void testVersionlessManagedDependency()
 
         ProjectBuildingException e = assertThrows( ProjectBuildingException.class,
                       () -> getContainer().lookup( org.apache.maven.project.ProjectBuilder.class ).build( pomFile, configuration ) );
-        assertThat( e.getMessage(),
-                    containsString( "[ERROR] 'dependencies.dependency.version' for org.apache.maven.its:a:jar is missing. "
-                        + "@ line 9, column 17" ) );
+        assertThat( e.getResults(), contains( projectBuildingResultWithProblemMessage(
+                "'dependencies.dependency.version' for org.apache.maven.its:a:jar is missing" ) ) );
+        assertThat( e.getResults(), contains( projectBuildingResultWithLocation( 17, 9) ) );
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/project/ProjectBuilderTest.java:111]: Trailing space missing

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithLocationMatcher.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        final ProjectBuildingResult r = (ProjectBuildingResult) o;
+
+        return r.getProblems().stream()
+                .anyMatch( p -> p.getLineNumber() == lineNumber && p.getColumnNumber() == columnNumber );
+    }
+
+    @Override
+    public void describeTo( Description description )
+    {
+        description.appendText("a ProjectBuildingResult with location ")
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithLocationMatcher.java:58]: space

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        final ProjectBuildingResult r = (ProjectBuildingResult) o;
+
+        return r.getProblems().stream()
+                .anyMatch( p -> p.getMessage().contains( problemMessage ) );
+    }
+
+    @Override
+    public void describeTo( Description description )
+    {
+        description.appendText("a ProjectBuildingResult with message ")
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java:56]: space

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java (hunk 4 above)
Comment (mthmulders) [reply maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java]: Fixed.

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuilderTest.java (hunk 2 above)
Comment (mthmulders) [reply maven-core/src/test/java/org/apache/maven/project/ProjectBuilderTest.java]: Fixed.

File: maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java (hunk 1 above)
Comment (mthmulders) [reply maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java:70]: That's what the original message said (see line 126 in the old code), it's a literal move.

Comment (michael-o) [reply maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java:70]: I guess the message applies to the entire reactor, doesn't it? If so, leave as-is.

Comment (mthmulders) [reply maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java:70]: > I guess the message applies to the entire reactor, doesn't it?

It does. I'll consider this resolved.

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithLocationMatcher.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * Hamcrest matcher to help create fluent assertions about {@link ProjectBuildingResult} instances.
+ */
+class ProjectBuildingResultWithLocationMatcher extends BaseMatcher<ProjectBuildingResult>
+{
+    private final int columnNumber;
+    private final int lineNumber;
+
+    ProjectBuildingResultWithLocationMatcher( int columnNumber, int lineNumber ) {
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithLocationMatcher.java:37]: curly on new line

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithLocationMatcher.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        this.columnNumber = columnNumber;
+        this.lineNumber = lineNumber;
+    }
+
+    @Override
+    public boolean matches( Object o )
+    {
+        if ( !( o instanceof ProjectBuildingResult ) ) {
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithLocationMatcher.java:45]: curly on new line

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithLocationMatcher.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            final ProjectBuildingResult r = (ProjectBuildingResult) o;
+            description.appendText( "was a ProjectBuildingResult with locations " );
+            String messages = r.getProblems().stream()
+                    .map( p -> formatLocation( p.getColumnNumber(), p.getLineNumber() ) )
+                    .collect( joining( ", ") );
+            description.appendText( messages );
+        }
+    }
+
+    static Matcher<ProjectBuildingResult> projectBuildingResultWithLocation( int columnNumber, int lineNumber ) {
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithLocationMatcher.java:84]: curly on new line

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+import static java.util.stream.Collectors.joining;
+
+/**
+ * Hamcrest matcher to help create fluent assertions about {@link ProjectBuildingResult} instances.
+ */
+class ProjectBuildingResultWithProblemMessageMatcher extends BaseMatcher<ProjectBuildingResult>
+{
+    private final String problemMessage;
+
+    ProjectBuildingResultWithProblemMessageMatcher( String problemMessage ) {
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java:36]: curly on new line

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        this.problemMessage = problemMessage;
+    }
+
+    @Override
+    public boolean matches( Object o )
+    {
+        if ( !( o instanceof ProjectBuildingResult ) ) {
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java:43]: curly on new line

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    public void describeTo( Description description )
+    {
+        description.appendText( "a ProjectBuildingResult with message " )
+                .appendValue(problemMessage);
+    }
+
+    @Override
+    public void describeMismatch(final Object o, final Description description)
+    {
+        if ( !( o instanceof ProjectBuildingResult ) ) {
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java:63]: curly on new line

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+            description.appendText( "was a ProjectBuildingResult with messages " );
+            String messages = r.getProblems().stream()
+                    .map( ModelProblem::getMessage )
+                    .map( m -> "\"" + m + "\"" + System.lineSeparator() )
+                    .collect( joining( ", ") );
+            description.appendText( messages );
+        }
+    }
+
+    static Matcher<ProjectBuildingResult> projectBuildingResultWithProblemMessage( String message ) {
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/project/ProjectBuildingResultWithProblemMessageMatcher.java:78]: curly on new line

**Review by michael-o — CHANGES_REQUESTED** [review]

**Comment by mthmulders** [comment]
Apologies for the curly on new line noise... Should've seen that myself!

**Comment by michael-o** [comment]
> Apologies for the curly on new line noise... Should've seen that myself!

No issue...

**Review by michael-o — APPROVED** [review]
Diffed two build of Core ITs. A very nice improvement! Please squash and merge.

**Comment by mthmulders** [comment]
> A very nice improvement!

Thanks! :-)

**Comment by jira-importer** [comment]
Resolve #9294

**Pull request closed without merging**